### PR TITLE
[jax2tf] First draft of top_k conversion to tf.

### PR DESCRIPTION
**Pull request by bchetioui**
We are able to use tf.math.top_k in most cases but:
- bfloat16 and complex numbers are unsupported
- tf.math.top_k works properly with args of type int32/64, uint32/64 and
  float32/64. Other flavors of uints/ints and bool can not be used with
  the operation by default, and they are thus promoted to their
  equivalent 32 bit representation, and the result is casted back to the
  original dtype.

A limitation right now is the handling of edge cases: TF and JAX order
np.inf and np.nan differently, resulting in inconsistencies in the case
where these values are present.

### CODE REVIEW

File: jax/experimental/jax2tf/jax2tf.py [hunk 1]
```diff
@@ -1113,6 +1113,29 @@ def func1(*jax_args):
...
+  # to a compatible type without loss of precision.
+  def promote_dtype(dtype):
+    if dtype in [jnp.bool_, jnp.uint8, jnp.uint16]:
+      return jnp.uint32
+    if dtype in [jnp.int8, jnp.int16]:
+      return jnp.int32
+    if dtype in [jnp.float16]:
+      return jnp.float32
+    if dtype in [jnp.bfloat16]:
+      return tf.bfloat16
```
Comment (gnecula) [inline jax/experimental/jax2tf/jax2tf.py:1127]: I am wondering why we return here `tf.bfloat16`, and in all other branches we returns a `jnp` type?

File: jax/experimental/jax2tf/jax2tf.py [hunk 2]
```diff
@@ -1113,6 +1113,29 @@ def func1(*jax_args):
 
 tf_impl[lax.scan_p] = _scan
 
+def _top_k(operand: TfVal, k: int) -> Tuple[TfVal, TfVal]:
+  # Some types originally incompatible with tf.math.top_k can be promoted
+  # to a compatible type without loss of precision.
+  def promote_dtype(dtype):
```
Comment (gnecula) [inline jax/experimental/jax2tf/jax2tf.py:1119]: It seems that `dtype` here is a TF dtype (result of `operand.dtype`). If that is true, then it would be clearer if we compare `dtype` with `numpy` types, not `jax.numpy` types (e.g., `dtype in [np.bool_, np.uint8, ...]`)

File: jax/experimental/jax2tf/tests/primitives_test.py [hunk 3]
```diff
@@ -109,6 +109,24 @@ def test_pad(self, harness: primitive_harness.Harness):
       raise unittest.SkipTest("pad with negative pad not supported")
     self.ConvertAndCompare(harness.dyn_fun, *harness.dyn_args_maker(self.rng()))
 
+  @primitive_harness.parameterized(primitive_harness.lax_top_k)
+  def test_top_k(self, harness: primitive_harness.Harness):
+    if (harness.params["k"] > harness.params["shape"][-1] or
```
Comment (gnecula) [inline jax/experimental/jax2tf/tests/primitives_test.py:114]: I do not think that we need to test this here. The goal is to test the conversion of primitive usage (lax.top_k_p.bind). The negative `k` will be rejected before we even get to use the primitive.

File: jax/experimental/jax2tf/tests/primitive_harness.py [hunk 4]
```diff
@@ -319,6 +319,40 @@ def parameterized(harness_group: Iterable[Harness],
...
+lax_top_k = tuple( # random testing
+  Harness(f"_inshape={jtu.format_shape_dtype_string(shape, dtype)}_k={k}",
+          lax.top_k,
+          [RandArg(shape, dtype), StaticArg(k)],
+          shape=shape,
+          dtype=dtype,
+          k=k)
+  for dtype in jtu.dtypes.all #[np.float32, np.int32, np.uint32]
+  for shape in [(3,), (5, 3)]
+  for k in [-1, 1, 3, 4]
```
Comment (gnecula) [inline jax/experimental/jax2tf/tests/primitive_harness.py:331]: I do not think that we need to include invalid values for `k` here. See comment below.

File: jax/experimental/jax2tf/tests/primitive_harness.py [hunk 5]
```diff
@@ -319,6 +319,40 @@ def parameterized(harness_group: Iterable[Harness],
...
 )
 
+lax_top_k = tuple( # random testing
+  Harness(f"_inshape={jtu.format_shape_dtype_string(shape, dtype)}_k={k}",
+          lax.top_k,
+          [RandArg(shape, dtype), StaticArg(k)],
+          shape=shape,
+          dtype=dtype,
+          k=k)
+  for dtype in jtu.dtypes.all #[np.float32, np.int32, np.uint32]
```
Comment (gnecula) [inline jax/experimental/jax2tf/tests/primitive_harness.py:329]: Can we remove the comment at the end of the line?

File: jax/experimental/jax2tf/tests/primitive_harness.py [hunk 6]
```diff
@@ -319,6 +319,40 @@ def parameterized(harness_group: Iterable[Harness],
...
+  for shape in [(3,), (5, 3)]
+  for k in [-1, 1, 3, 4]
+  for rng_factory in [jtu.rand_default]
+) + tuple( # stability test
+  Harness(f"stability_inshape={jtu.format_shape_dtype_string(arr.shape, arr.dtype)}_k={k}",
```
Comment (gnecula) [inline jax/experimental/jax2tf/tests/primitive_harness.py:334]: I am curious how is stability involved in top_k? Is it about testing cases when we have repeated values?

Comment (SIben) [reply jax/experimental/jax2tf/tests/primitive_harness.py:334]: `top_k` returns the top k elements as well as their indices within the array; for that reason, it is important to know how the case in which two elements are the same is treated. The [TensorFlow docs](https://www.tensorflow.org/api_docs/python/tf/math/top_k) mentions that "If two elements are equal, the lower-index element appears first.", hence the stability check.

File: jax/experimental/jax2tf/tests/primitives_test.py (hunk 3 above)
Comment (SIben) [reply jax/experimental/jax2tf/tests/primitives_test.py:114]: I believe that we should keep these tests to make sure `lax.top_k` always fails in potential edge cases. Because the clamping behavior is not always consistent, it is important that we know if that invalid parameters are never silently passed to the function.

File: jax/experimental/jax2tf/tests/primitive_harness.py (hunk 4 above)
Comment (SIben) [reply jax/experimental/jax2tf/tests/primitive_harness.py:331]: See argument below.

File: jax/experimental/jax2tf/jax2tf.py (hunk 1 above)
Comment (SIben) [reply jax/experimental/jax2tf/jax2tf.py]: Addressed in latest commits.

File: jax/experimental/jax2tf/jax2tf.py (hunk 2 above)
Comment (SIben) [reply jax/experimental/jax2tf/jax2tf.py]: Addressed in latest commits.

File: jax/experimental/jax2tf/tests/primitive_harness.py (hunk 5 above)
Comment (SIben) [reply jax/experimental/jax2tf/tests/primitive_harness.py]: Addressed in latest commits.

**Review by gnecula — APPROVED** [review]

**Comment by gnecula** [comment]
I tested this internally before merging.
